pex/pip/tailer.py

```diff
@@ -0,0 +1,137 @@
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import absolute_import, print_function
+
+import errno
+import os
+import re
+from threading import Condition, Event, Thread
+from types import TracebackType
+
+from pex.compatibility import get_stdout_bytes_buffer
+from pex.typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from typing import Any, BinaryIO, Iterable, Optional, Pattern, Type
+
+
+class Tailer(Thread):
+    @classmethod
+    def tail(
+        cls,
+        path,  # type: str
+        encoding="utf-8",  # type: str
+        filters=(),  # type: Iterable[Pattern]
+        poll_interval=0.1,  # type: float
+        output=get_stdout_bytes_buffer(),  # type: BinaryIO
+    ):
+        # type: (...) -> Tailer
+        tailer = cls(
+            path, encoding=encoding, filters=filters, poll_interval=poll_interval, output=output
+        )
+        tailer.start()
+        return tailer
+
+    def __init__(
+        self,
+        path,  # type: str
+        encoding="utf-8",  # type: str
+        filters=(),  # type: Iterable[Pattern]
+        poll_interval=0.1,  # type: float
+        output=get_stdout_bytes_buffer(),  # type: BinaryIO
+    ):
+        # type: (...) -> None
+        super(Tailer, self).__init__(name="Tailing {path}".format(path=path))
+        self.daemon = True
+
+        self._path = path
+        self._encoding = encoding
+        self._filters = filters or [re.compile(r".*")]
+        self._poll_interval = poll_interval
+        self._output = output
+
+        self._offset = 0
+        self._observer = Condition()
+        self._stopped = Event()
+
+    def __enter__(self):
+        # type: () -> Tailer
+        if not self._stopped.is_set() and not self.is_alive():
+            self.start()
+        return self
+
+    def __exit__(
+        self,
+        _exc_type,  # type: Optional[Type]
+        _exc_val,  # type: Optional[Any]
+        _exc_tb,  # type: Optional[TracebackType]
+    ):
+        # type: (...) -> None
+        self.stop()
+
+    def run(self):
+        # type: () -> None
+        while not self._stopped.is_set():
+            if not self._tail():
+                self._stopped.wait(self._poll_interval)
+        self._tail()
+
+    def _tail(self):
+        # type: () -> bool
+        try:
+            offset = os.path.getsize(self._path)
+            if offset <= self._offset:
+                # We do not handle generic tailing here where the file may be truncated and then
+                # re-grow. The Pip log case is a simple always growing file.
+                return False
+        except OSError as e:
+            if e.errno == errno.ENOENT:
+                # The file may not exist yet; so we just wait for it to appear.
+                return False
+            raise e
+
+        with open(self._path, "rb") as fp:
+            fp.seek(self._offset)
+            while fp.tell() < offset:
+                line_bytes = fp.readline()
```

Comment:
Correct. But since the file is being written as streaming, every time this method runs it will encounter EOF. But it might encounter an EOF in the middle of a line if the writer has written half a line at the point when you read.

Reply:
Ah, yes. This needs to be fixed up, or commented if it doesn't happen to actually need fixup.